pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/RoutingManager.java

```diff
@@ -0,0 +1,481 @@
+   */
+  public synchronized void refreshSegment(String tableNameWithType, String segment) {
+    LOGGER.info("Refreshing segment: {} for table: {}", segment, tableNameWithType);
+    RoutingEntry routingEntry = _routingEntryMap.get(tableNameWithType);
+    if (routingEntry != null) {
+      routingEntry.refreshSegment(segment);
+      LOGGER.info("Refreshed segment: {} for table: {}", segment, tableNameWithType);
+    } else {
+      LOGGER.warn("Routing does not exist for table: {}, skipping refreshing segment", tableNameWithType);
+    }
+  }
+
+  /**
+   * Returns {@code true} if the routing exists for the given table.
+   */
+  public boolean routingExists(String tableNameWithType) {
+    return _routingEntryMap.containsKey(tableNameWithType);
+  }
+
+  /**
+   * Returns the routing table (map from server instance to list of segments hosted by the server) based on the broker
+   * request.
+   * <p>NOTE: The broker request should already have the table suffix (_OFFLINE or _REALTIME) appended.
+   */
+  public Map<ServerInstance, List<String>> getRoutingTable(BrokerRequest brokerRequest) {
+    String tableNameWithType = brokerRequest.getQuerySource().getTableName();
+    RoutingEntry routingEntry = _routingEntryMap.get(tableNameWithType);
+    Preconditions.checkState(routingEntry != null, "Failed to find routing for table: %s", tableNameWithType);
+
+    Map<String, String> segmentToInstanceMap = routingEntry.calculateSegmentToInstanceMap(brokerRequest);
+    Map<ServerInstance, List<String>> routingTable = new HashMap<>();
+    for (Map.Entry<String, String> entry : segmentToInstanceMap.entrySet()) {
+      ServerInstance serverInstance = _enabledServerInstanceMap.get(entry.getValue());
+      if (serverInstance != null) {
+        routingTable.computeIfAbsent(serverInstance, k -> new ArrayList<>()).add(entry.getKey());
+      } else {
+        // Should not happen in normal case unless encountered unexpected exception when updating routing entries
+        _brokerMetrics.addMeteredTableValue(tableNameWithType, BrokerMeter.SERVER_MISSING_FOR_ROUTING, 1L);
+      }
+    }
+    return routingTable;
+  }
+
+  /**
+   * Returns the time boundary info for the given offline table name.
+   * <p>NOTE: Time boundary info is only available for the offline part of the hybrid table.
+   */
+  @Nullable
+  public TimeBoundaryInfo getTimeBoundaryInfo(String offlineTableName) {
+    RoutingEntry routingEntry = _routingEntryMap.get(offlineTableName);
+    Preconditions.checkState(routingEntry != null, "Failed to find routing for table: %s", offlineTableName);
+    TimeBoundaryManager timeBoundaryManager = routingEntry.getTimeBoundaryManager();
+    return timeBoundaryManager != null ? timeBoundaryManager.getTimeBoundaryInfo() : null;
+  }
+
+  private static class RoutingEntry {
+    final String _tableNameWithType;
+    final SegmentSelector _segmentSelector;
+    final SegmentPruner _segmentPruner;
```

Comment:
We should consider to make this to `List<SegmentPruner>`.

For instance, it's a valid use case to do the pruning on both partition and time column value.

Reply:
Good point. I was planning to add this later when more segment pruners are implemented, but I can also add them now to avoid unnecessary config change in the future

Reply:
Added